### PR TITLE
[doc] Testplans are now generated as markdown.

### PR DESCRIPTION
**Pull request by HU90m**
Testplans are now rendered into a markdown document as oppose to a large HTML table.

These HTML tables were unwieldy when viewed from the site or as plain text. The original `pipe` output was not standard markdown.

New version of the [GPIO Testplan](https://opentitan.org/book/hw/ip/gpio/data/gpio_testplan.html):

image

Related to https://github.com/lowRISC/opentitan/issues/19078

### CODE REVIEW

File: util/dvsim/testplanner.py [hunk 1]
```diff
@@ -36,9 +36,8 @@ def main():
         if args.sim_results:
             outfile.write(
                 testplan.get_sim_results(args.sim_results, fmt="html"))
-
         else:
-            outfile.write(testplan.get_testplan_table("html"))
+            testplan.write_testplan_doc(outfile)
```
[the remark this answers is not in the document]
Comment (HU90m) [reply util/dvsim/testplanner.py:40]: In the codebase, no other scripts or tools use it, and I'm not aware of anywhere it is used outside of the documentation build.

**Comment by pamaury** [comment]
That's much more readable, thanks Hugo

**Review by jwnrt — APPROVED** [review]

File: util/dvsim/Testplan.py [hunk 2]
```diff
@@ -481,73 +482,35 @@ def get_stage_regressions(self):
...
-                table.append([tp.stage, tp.name, tests, desc])
-            lines += [
-                tabulate(table,
-                         headers=header,
-                         tablefmt=tabfmt,
-                         colalign=colalign)
-            ]
+        stages = {}
+        for tp in self.testpoints:
+            try:
```
Comment (pamaury) [inline util/dvsim/Testplan.py:490]: I would find it more clearer if `if t.stage not in stage` but that's just a personal preference.

**Review by pamaury — APPROVED** [review]
Looks good to me.

File: util/dvsim/Testplan.py (hunk 2 above)
Comment (rswarbrick) [reply util/dvsim/Testplan.py:490]: I think that something like the following is the approach that I've usually seen:
```python3
        stages = {}
        for tp in self.testpoints:
            stages.setdefault(tp.stage, list()).append(tp)
```

Comment (HU90m) [reply util/dvsim/Testplan.py:490]: ooooo `setdefault` is very nice

Comment (pamaury) [reply util/dvsim/Testplan.py:490]: Didn't know `setdefault`, will definitely use it in the future! Thanks @rswarbrick

**Comment by HU90m** [comment]
Now using `setdefault`

File: util/dvsim/Testplan.py [hunk 3]
```diff
@@ -481,73 +482,32 @@ def get_stage_regressions(self):
...
-                tabulate(table,
-                         headers=header,
-                         tablefmt=tabfmt,
-                         colalign=colalign)
-            ]
+        stages = {}
+        for tp in self.testpoints:
+            stages.setdefault(tp.stage, list()).append(tp)
+
+        output.write("# Testplan\n## Testpoints\n")
```
Comment (marnovandermaas) [inline util/dvsim/Testplan.py:492]: Do we usually have two enters after a heading?

**Review by marnovandermaas — APPROVED** [review]
Thanks for implementing this!

**Comment by HU90m** [comment]
@hcallahan-lowrisc Fixed the md headings in testplan problem.

**Comment by HU90m** [comment]
Added double newline after headings

**Comment by jwnrt** [comment]
Thanks @HU90m

Will these files be included with `CMDGEN` in the future?

**Comment by HU90m** [comment]
> Will these files be included with `CMDGEN` in the future?

That's the plan

**Comment by hcallahan-lowrisc** [comment]
> > Will these files be included with `CMDGEN` in the future?
> 
> That's the plan

Are we sure about this? Wouldn't that just be duplicating the content from the .hjson? I think that could be confusing.

**Comment by HU90m** [comment]
> Are we sure about this? Wouldn't that just be duplicating the content from the .hjson? I think that could be confusing.

Could this not be said about the register docs as well?

**Comment by hcallahan-lowrisc** [comment]
> > Are we sure about this? Wouldn't that just be duplicating the content from the .hjson? I think that could be confusing.
> 
> Could this not be said about the register docs as well?

Partly yes, but there is a transformation to add wavejson and covert fields to tables. With the testplans, there really isn't much being transformed.

**Pull request merged**